### DISCUSSION TITLE
How to modify metadata only

### DISCUSSION THREAD

**Discussion by Enablator** [minio/minio#11545] [Q&A] [answered]
My flow is the following
1) download an object from s3
2) perform some analysis
3) store analysis results as an object metadata
Is it possible just to update metadata of the object without uploading the whole file back? The reason I want it is because I'm working with big files e.g. 5gb. So I don't want to download and then to upload the whole file

**Comment by harshavardhana** [comment, marked as answer]
You can use server-side CopyObject() to update metadata https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html

**Reply by Enablator** [reply]
I checked minioClient methods and there is only one implementation of a copy logic - CopyObject method. 
And it works but only for files less than 5 gb.  

For bigger files server side multipart upload should be used according to aws documentation. 
But in minioClient the only method which has similar multipart logic is composeObject. And at the same time composeObject method doesn't populate necessary for copying headers. So aws throws an error. 
I tried to manually add header "x-amz-metadata-directive":"REPLACE" and it seems to be working but it's a sort of a workaround for composeObject method.

Do you have a plan to implement such multipart copying logic in minioClient?

**Reply by balamurugana** [reply]
@Enablator Are you using `minio-java`? and where did you have `x-amz-metadata-directive` header? Logically speaking, `CreateMultipartUpload` S3 API does not have such header and you don't need to pass the header. As it is `REPLACE` which is however a default behavior i.e. it is not possible to `COPY` metadata.

**Reply by Enablator** [reply]
@balamurugana ,
I'm using minio-java. 
For small files I used copyObject method. I was adding a new header but got an error from s3 about copying of a file into itself without modifying tags/headers etc. Then I used method in args builder to set metadataDirective. It worked. After that copying worked.
But for bigger than 5GB files I didn't find an appropriate multipart copy method in minio-java. But there is a composeObject method which uses multipart server-side upload.
So I used composeObject method in similar to copyObject way. I added a new custom header with metadata prefix. But it didn't work. I was getting an error from s3 that i'm trying to copy object into itself without modifying tags/metadata etc. 
After that I added one more header "x-amz-metadata-directive". And it worked. Big object was copied using multipart upload into itself within composeObject method. But this is a workaround. 

I think there should be a logic inside of 'copyObject' method which initiate MultipartUpload when file size is bigger than 5GB.

**Reply by balamurugana** [reply]
@Enablator Could you enable `MinioClient.traceOn()` for workaround case and share the output for > 5GB file?

**Comment by klauspost** [comment]
https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html

**Reply by Enablator** [reply]
Thanks for the idea. I will check if tags will work for me

**Reply by Enablator** [reply]
@klauspost thank you! 
I will use tags!
